torch/distributed/tensor/_ops/single_dim_strategy.py

```diff
@@ -178,49 +179,133 @@ def _expand_single_dim_strategy_to_mesh(
+                op, args_schema, kwargs_schema
+            )
+            strategies_over_one_mesh_dim = _insert_single_dim_replication_strategy(
+                strategies_over_one_mesh_dim, num_inputs
+            )
+            expanded_strategies_over_one_mesh_dim = (
+                _fill_single_dim_strategy_placeholders(
+                    unique_input_placements, strategies_over_one_mesh_dim
+                )
+            )
+
+            # Note: does not support `allow_unbacked_sharding` which is needed by matmul rules for some compile test
+            # currently, we should probably change that test though, since it seems wrong to me to allow sharding unbacked
+            # dims
+            return expand_to_full_mesh_op_strategy(
+                mesh,
+                op_schema,
+                cast(list[PlacementList], expanded_strategies_over_one_mesh_dim),
+            )
+
+        return expanded_strategy
+
+    def _translate_foreach_op_schema(op_schema: OpSchema, index: int) -> OpSchema:
+        """Translate foreach op to per-element version of schema."""
+        op_parts = str(op_schema.op).split(".")
+        base_op_name = op_parts[-2].replace("_foreach_", "")
+        foreach_variant = op_parts[-1]
+
+        # select per-element args
+        target_args = []
+        for arg in op_schema.args_schema:
+            if isinstance(arg, TupleStrategy):
+                target_args.append(arg.children[index])
+            else:  # single arg instead of list
+                target_args.append(arg)
+
+        # figure out target op variant
+        variant_map = {
+            "List": "Tensor",
+            "ScalarList": "Scalar",
+            "Scalar": "Scalar",
+            "Tensor": "Tensor",
+            "default": "default",
+        }
+        target_variant = (
+            "default"
+            if len(target_args) == 1
+            else variant_map.get(foreach_variant, "default")
         )
-        expanded_strategies_over_one_mesh_dim = _fill_single_dim_strategy_placeholders(
-            unique_input_placements, strategies_over_one_mesh_dim
+
+        # this seems a bit messy
+        base_op = getattr(torch.ops.aten, base_op_name)
+        target_op = (
+            getattr(base_op, target_variant)
+            if target_variant in base_op.overloads()
+            else base_op.default
         )
 
-        # Note: does not support `allow_unbacked_sharding` which is needed by matmul rules for some compile test
-        # currently, we should probably change that test though, since it seems wrong to me to allow sharding unbacked
-        # dims
-        return expand_to_full_mesh_op_strategy(
-            mesh,
-            op_schema,
-            cast(list[PlacementList], expanded_strategies_over_one_mesh_dim),
+        return OpSchema(
+            target_op,  # type: ignore[arg-type]
+            args_schema=tuple(target_args),
+            kwargs_schema=op_schema.kwargs_schema,
```

Comment:
note: I don't know of if there are any cases where kwargs_schema could contain a TupleStrategy and that would need to also be translated.